File: core/src/main/scala/eu/easyminer/rdf/data/TriplesOps.scala
package eu.easyminer.rdf.data

import eu.easyminer.rdf.data.Triple.TripleTraversableView

/**
  * Created by Vaclav Zeman on 2. 2. 2018.
  */
trait TriplesOps {

  def toTriples: TripleTraversableView

  def histogram(s: Boolean, p: Boolean, o: Boolean): Histogram = {
    def boolToOpt[T](x: T, bool: Boolean) = if (s) Some(x) else None

    def tripleToKey(triple: Triple) = Histogram.Key(
      boolToOpt(triple.subject, s),
      boolToOpt(triple.predicate, p),
      boolToOpt(triple.`object`, o)
    )

    Histogram(toTriples.map(tripleToKey))
  }

  def types(): Vector[PredicateInfo] = PredicateInfo(toTriples)

}

File: java/src/main/java/com/github/propi/rdfrules/java/TriplesOps.java
package com.github.propi.rdfrules.java;

import java.util.Map;

/**
 * Created by Vaclav Zeman on 2. 5. 2018.
 */
public interface TriplesOps {

    com.github.propi.rdfrules.data.ops.TriplesOps asScala();

    default Map<HistogramKey, Integer> histogram(boolean subject, boolean predicate, boolean object) {
        return new MapWrapper<>(asScala().histogram(subject, predicate, object).getMap())
                .mapValues(x -> ((Integer) x))
                .mapKeys(HistogramKey::new, HistogramKey::asScala)
                .asJava();
    }

    default Iterable<PredicateInfo> types() {
        return new IterableWrapper<>(asScala().types().toIterable()).map(PredicateInfo::new).asJava();
    }

}